### PR TITLE
DPL-823 [Part 2 - upload] Import PBMC pool plates into Sequencescape

### LINKED ISSUE #3857: DPL-823 Import PBMC pool plates into Sequencescape
**User story**
As a member of the SSR team, who has received a plate from faculty containing thawed, pooled PBMCs, I would like to be able to import them into Sequencescape along with all the required data.

**Who are the primary contacts for this story**
Lesley, Abby, Katy, Liz H

**Who is the nominated tester for UAT**
Abby, Liz H

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] I can upload the plate as a "LRC PBMC Pools Input" plate in Sequencescape
- [ ] Record pools, with the following information:
    - [ ] The samples that make up the pool
- [ ] For each sample, record:
    - [ ] ~Study~ --> currently has to be the same across the whole manifest
    - [ ] Donor id
    - [ ] Supplier sample name (Vacutainer barcode)
    - [ ] Which well it's in
    - [ ] Unique tag_depth index (assign automatically on upload, not provided by user), to avoid tag clash and enable compound sample creation when sequencing request is made later on
- [ ] Consider whether to insert rows into the stock resource table in the MLWH or not.
- [ ] Update [documentation](https://ssg-confluence.internal.sanger.ac.uk/display/PSDPUB/scRNA+Core+Pipeline+Documentation) - it currently has instructions from DPL-809 (linked below) - add similar instructions regarding this import.

**References**
This story has a non-blocking relationship with:
- https://github.com/sanger/sequencescape/issues/3849 - equivalent but for fresh blood
- https://github.com/sanger/sequencescape/issues/3937 - equivalent but for cDNA

**Additional information**
This will use the normal manifest process - SSRs generate the manifest & barcode labels, send them to faculty, who use the barcodes on their plates and send the manifest back filled in. 

This is new functionality for manifests - existing manifests that import pools (e.g. multiplexed library manifests) do not include the individual sample information.

### PR DESCRIPTION
**Pull request by KatyTaylor**
Closes #3857 

#### Changes proposed in this pull request

- Enable 'rows_per_well' feature for manifest generation from https://github.com/sanger/sequencescape/pull/4017 
- Allow upload of pools plates
    - Only thing that needed sorting out was the 'tag clash' when multiple untagged samples / aliquots were saved into the same well
    - To allow this, we set a unique 'tag_depth' on each aliquot in the well
    - This will also be needed for later in the pipeline when these pools become 'compound samples' prior to sequencing and prior to being exported to the warehouse
- It nicely ignores rows that haven't been filled in, so the manifest is generated with excess rows per pool and not all have to be filled
- Added tests in sample_manifest_spec and uploader_spec

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

### CODE REVIEW

File: spec/models/sample_manifest/uploader_spec.rb [hunk 1]
```diff
@@ -302,5 +304,48 @@
...
+
+          wells.each do |well|
+            aliquots = well.aliquots
+
+            # within a well, each aliquot should have its own tag_depth
+            expect(aliquots.map(&:tag_depth).uniq.count).to eq(aliquots.count)
+          end
+        end
+      end
+    end
```
Comment (andrewsparkes) [inline spec/models/sample_manifest/uploader_spec.rb:349]: Maybe add tests for where a file is uploaded and some and/or all of the rows for a particular well are not filled in, to check that it works ok and correctly filters out the blanks without losing the rows that do have info.
e.g. num_rows_per_well 3, and A1 has 3 rows, but only 2 are filled in. or A2 has 3 rows and none are filled in.
(Or maybe this is already tested elsewhere as more basic functionality of filtering out empty rows first before we get here?)

**Review by andrewsparkes — APPROVED** [review]
Some discussion comments but looks good

File: app/models/sample_manifest.rb [hunk 2]
```diff
@@ -129,11 +129,32 @@ def default_filename
     "#{study_id}stdy_manifest_#{id}_#{created_at.to_formatted_s(:dmy)}"
   end
 
-  # Use a default value of 1 for rows_per_well if not set
+  # Number of rows per well in the manifest file, specified in manifest_types.yml.
```
[the remark this answers is not in the document]
Comment (KatyTaylor) [reply app/models/sample_manifest.rb:132]: It's a good point - I have it on my to do list to ask what the number should be. It kind of has to be high enough, because the spreadsheets are locked down and won't let them insert new rows. This is for a good reason, because some data is created upfront on generation of the manifest (Sanger Sample Ids, Sample Manifest Assets), so if they could insert rows then this data wouldn't exist for them.

File: spec/models/sample_manifest/uploader_spec.rb (hunk 1 above)
Comment (KatyTaylor) [reply spec/models/sample_manifest/uploader_spec.rb:352]: OK I've had a look and I think it's alright. There's a test in this file `'will upload a valid partial plate sample manifest'` which only fills out some of the rows. There's also unit testing of the specific bit of code that ignores empty rows - the method called `empty?` in `row.rb` is tested in `row_spec.rb`, test called `'knows if it is empty'`. Think that's enough as I didn't have to add any additional functionality around this.
